src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -0,0 +1,267 @@
+    pub(crate) async fn contains_file(&self, path: &str) -> bool {
+        self.mem_cache.contains_key(path)
+            && self.file_cache.stat(path, OpStat::default()).await.is_ok()
+    }
+
+    /// Read from a specific path using the OpRead operation.
+    /// It will attempt to retrieve the data from the local cache.
+    /// If the data is not found in the local cache,
+    /// it will fallback to retrieving it from remote object storage
+    /// and cache the result locally.
+    pub(crate) async fn read<'life0, 'life1, 'async_trait, I, F>(
+        &'life0 self,
+        path: &'life1 str,
+        args: OpRead,
+        inner_read: F,
+    ) -> Result<(RpRead, Box<dyn Read>)>
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        I: Accessor,
+        F: FnOnce(
+            &'life1 str,
+            OpRead,
+        ) -> Pin<
+            Box<dyn Future<Output = Result<(RpRead, I::Reader)>> + Send + 'async_trait>,
+        >,
+    {
+        let read_key = read_cache_key(path, &args);
+        let read_result = self.mem_cache.get(&read_key).await;
+
+        if let Some(read_result) = read_result {
+            let cache_result = match read_result {
+                ReadResult::Success(_) => {
+                    increment_counter!(OBJECT_STORE_LRU_CACHE_HIT, "result" => "success");
+
+                    // There is a concurrent issue here, the local cache may be purged
+                    // while reading, we have to fallback to remote read
+                    match self.file_cache.read(&read_key, OpRead::default()).await {
+                        Ok(ret) => Ok(to_output_reader(ret)),
+                        Err(_) => inner_read(path, args.clone()).await.map(to_output_reader),
+                    }
+                }
+                ReadResult::NotFound => {
+                    increment_counter!(OBJECT_STORE_LRU_CACHE_HIT, "result" => "not_found");
+
+                    Err(OpendalError::new(
+                        ErrorKind::NotFound,
+                        &format!("File not found: {path}"),
+                    ))
+                }
+            };
+            self.mem_cache.run_pending_tasks().await;
+
+            return cache_result;
+        }
+
+        increment_counter!(OBJECT_STORE_LRU_CACHE_MISS);
+
+        let inner_result = inner_read(path, args.clone()).await;
+        match inner_result {
+            Ok((rp, mut reader)) => {
+                let (_, mut writer) = self.file_cache.write(&read_key, OpWrite::new()).await?;
+
+                while let Some(bytes) = reader.next().await {
+                    writer.write(&bytes?).await?;
+                }
+
+                // Call `close` to ensure data is written.
+                writer.close().await?;
+
+                // Read from local cache immediately, because it may be purged while inserting.
+                let cache_result = self.file_cache.read(&read_key, OpRead::default()).await;
```

Comment:
Will this `read()` get a partial file when two "same" tasks run concurrently? One task closes the cache file, and another task starts writing its content. Then the first task will see the new partial file created by the second task

Reply:
It won't. We use atomic write(based on `rename`) for cache files too.

https://github.com/GreptimeTeam/greptimedb/blob/fe783c7c1f85c10bf57e8b3ef6e5fb9f19b8189e/src/datanode/src/store.rs#L115